tests/44account_data.pl

```diff
@@ -0,0 +1,269 @@
+push our @EXPORT, qw( matrix_add_account_data );
+
+=head2 matrix_add_account_data
+
+   matrix_add_account_data( $user, $type, $content )->get;
+
+Add account data for the user.
+
+=cut
+
+sub matrix_add_account_data
+{
+   my ( $user, $type, $content ) = @_;
+
+   do_request_json_for( $user,
+      method  => "PUT",
+      uri     => "/v2_alpha/user/:user_id/account_data/$type",
+      content => $content
+   );
+}
+
+=head2 matrix_add_room_account_data
+
+    matrix_add_account_data( $user, $room_id, $type, $content )->get;
+
+Add account data for the user for a room.
+
+=cut
+
+sub matrix_add_room_account_data
+{
+   my ( $user, $room_id, $type, $content ) = @_;
+
+   do_request_json_for( $user,
+      method  => "PUT",
+      uri     => "/v2_alpha/user/:user_id/rooms/$room_id/account_data/$type",
+      content => $content
+   );
+}
+
+
+test "Can add account data",
+   requires => [ local_user_fixture() ],
+
+   check => sub {
+      my ( $user ) = @_;
+
+      matrix_add_account_data( $user, "my.test.type", {} );
+   };
+
+
+test "Can add account data to room",
+   requires => [ local_user_and_room_fixtures() ],
+
+   check => sub {
+      my ( $user, $room_id ) = @_;
+
+      matrix_add_room_account_data( $user, $room_id, "my.test.type", {} );
+   };
+
+
+sub check_one_account_data_event
+{
+   my ( $account_data_events, $expected_type, $cats_or_rats ) = @_;
+
+   log_if_fail "account data", $account_data_events;
+
+   @{ $account_data_events } == 1 or die "Expected only one event";
+
+   my $event = $account_data_events->[0];
+   assert_json_keys($event, qw( type content ));
+
+   $event->{type} eq $expected_type
+      or die "Unexpected event type, wanted $expected_type";
+
+   $event->{content}{cats_or_rats} eq $cats_or_rats
+      or die "Unexpected event content, wanted $cats_or_rats";
+}
+
+
+sub setup_account_data
+{
+   my ( $user, $room_id ) = @_;
+
+   Future->needs_all(
+      matrix_add_account_data( $user, "my.test.type", {
+         cats_or_rats => "frogs",
+      }),
+      matrix_add_room_account_data( $user, $room_id, "my.test.type", {
+         cats_or_rats => "dogs",
+      }),
+   )->then( sub {
+      Future->needs_all(
+         matrix_add_account_data( $user, "my.test.type", {
+            cats_or_rats => "cats",
+         }),
+         matrix_add_room_account_data( $user, $room_id, "my.test.type", {
+            cats_or_rats => "rats",
+         }),
+      );
+   });
+}
+
+
+test "Latest account data comes down in /initialSync",
+   requires => [ local_user_and_room_fixtures() ],
+
+   check => sub {
+      my ( $user, $room_id ) = @_;
+
+      setup_account_data( $user, $room_id )->then( sub {
+         matrix_initialsync( $user );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         check_one_account_data_event(
+            $body->{account_data}, "my.test.type", "cats"
+         );
+
+         check_one_account_data_event(
+            $body->{rooms}[0]->{account_data}, "my.test.type", "rats"
```

Comment:
This inner arrow isn't required; as per STYLE-GUIDE it should be omitteed.

    $body->{rooms}[0]{account_data} ...